lib/withOnyx.d.ts

```diff
@@ -0,0 +1,159 @@
+import {IsEqual} from 'type-fest';
+import {CollectionKeyBase, KeyValueMapping, OnyxCollectionEntries, OnyxEntry, OnyxKey, Selector} from './types';
+
+/**
+ * Represents the base mapping options between an Onyx key and the component's prop.
+ */
+type BaseMapping<TComponentProps, TOnyxProps> = {
+    canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+    initWithStoredValues?: boolean;
+};
+
+/**
+ * Represents the string / function `key` mapping option between an Onyx key and the component's prop.
+ *
+ * If `key` is `string`, the type of the Onyx value that is associated with `key` must match with the type of the component's prop,
+ * otherwise an error will be thrown.
+ *
+ * If `key` is `function`, the return type of `key` function must be a valid Onyx key and the type of the Onyx value associated
+ * with `key` must match with the type of the component's prop, otherwise an error will be thrown.
+ *
+ * @example
+ * ```ts
+ * // Onyx prop with `string` key
+ * onyxProp: {
+ *     key: ONYXKEYS.ACCOUNT,
+ * },
+ *
+ * // Onyx prop with `function` key
+ * onyxProp: {
+ *     key: ({reportId}) => ONYXKEYS.ACCOUNT,
+ * },
+ * ```
+ */
+type BaseMappingKey<
+    TComponentProps,
+    TOnyxProps,
+    TOnyxProp extends keyof TOnyxProps,
+    TOnyxKey extends OnyxKey,
+    TOnyxValue
+> = IsEqual<TOnyxValue, TOnyxProps[TOnyxProp]> extends true
+    ? {
+          key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
+      }
+    : never;
+
+/**
+ * Represents the string `key` and `selector` mapping options between an Onyx key and the component's prop.
+ *
+ * The function signature and return type of `selector` must match with the type of the component's prop,
+ * otherwise an error will be thrown.
+ *
+ * @example
+ * ```ts
+ * // Onyx prop with `string` key and selector
+ * onyxProp: {
+ *     key: ONYXKEYS.ACCOUNT,
+ *     selector: (value: Account | null): string => value?.id ?? '',
+ * },
+ * ```
+ */
+type BaseMappingStringKeyAndSelector<
```

Comment:
can we combine `BaseMappingStringKeyAndSelector` and `BaseMappingFunctionKeyAndSelector`?

```ts
key: TOnyxKey | (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
```

Reply:
@hayata-suenaga @blazejkustra I'm leaving both separated because when they are combined TS end up losing some type-safety on `selector`.